### PR TITLE
Build a Twitch Chatbot in Node.js

### LINKED ISSUE #2411: Build a Twitch Chatbot in Node.js
### Proposed title of the article

Build a Twitch Chatbot in Node.js

### Introduction paragraph (2-3 paragraphs):

Twitch, to newcomers, is an online streaming platform that allows streamers and audiences to lively interact with each other. It begun as a small game streaming platform for gamers but soon got a leap ahead. It now not only supports other services but also contents from a broad spectrum. For instance cooking shows, live talk shows, daily lifestreaming, live chatting, and interactions among others. Streamers earn from monthly subscriptions or donations from fans depending on the content and as long as the content is within the [terms of service](https://www.twitch.tv/p/en/legal/terms-of-service/), [community guidelines](https://www.twitch.tv/p/en/legal/community-guidelines/) and [terms of sale](https://www.twitch.tv/p/en/legal/terms-of-sale/).

Due to the growth of the service, the need of interacting with the Twitch API to create more personalized services or functions arose. This led to the development and deployment of bots which interact with the service. The bots created by these genius developers range from chatting bots, viewers engagement bots, monitoring and statistical bots, polling bots among many others. Some of these top Twitch bots can be found [here](https://www.fairlyoddstreamers.com/resources/bots). Even though so many bots have been created or modified, still there is a need for personalized bots depending on the issue you as a developer or streamer want to handle.

In this tutorial, we will learn about the Twitch API and how to create a chatting bot in Node.js to interact with it.

### Key takeaways: 

At the end of the tutorial, you will learn:

- What is the Twitch API
- Key components, basics, and how to interact with the API in Node.js
- How to build a Twitch chatbot
- How to run a Twitch chatbot in Node.js

### Pre-requisites

Some of the basics needed for the tutorial include:

- Web development basics
- Little interactions with Twitch website. This helps you know how Twitch works and the interface you would be dealing with. If you are new, you can check it out at [https://www.twitch.tv/](https://www.twitch.tv/).
- An IDE installed on your machine. My recommendation is Visual Studio Code. You can get it [here](https://code.visualstudio.com/).
- Node environment installed on your machine. You can get it [here](https://nodejs.org/en/).
- A stable internet connection.

### References: 

- [Top Twitch bots](https://www.fairlyoddstreamers.com/resources/bots).
- [Twitch API documentation](https://www.npmjs.com/package/twitch).
- [Node.js documentation](https://nodejs.org/en/docs/meta/topics/dependencies/).

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by justusmbuvi**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**

This fixes #2411.
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by justusmbuvi** [comment]
This fixes #2411

**Comment by hectorkambow** [comment]
@justusmbuvi 
Good afternoon and thank you for submitting your article for publishing. It has been entered into our queue and should be reviewed (and published pending any changes) as soon as a peer reviewer is finished with the ones in the queue before it. 🚀

**Review by srishilesh — CHANGES_REQUESTED** [review]
Hey @justusmbuvi,

Welcome to EngEd community!

I'll be reviewing this PR.

Before I start the review, I would like a few changes to be done:
- Rename the directory `/Build-a-Twitch-Chatbot-in-Node.js/` to `/build-a-twitch-chatbot-in-nodejs/`
- Add a hero image of dimensions 600x400. Use Unsplash to download images, and Pixlr to edit them (if needed)
- Add authors page (headshot image + author bio). Have a look at `/contents/authors` to understand how its done. Or refer contribution guidelines to get started.

Feel free to ping me here or via Slack, in case of any questions.

Thanks.

**Comment by hectorkambow** [comment]
[2120420-3bb0dc0c2c5d4e1f8480-1625592568154.pdf](https://github.com/section-engineering-education/engineering-education/files/6771739/2120420-3bb0dc0c2c5d4e1f8480-1625592568154.pdf)


Plagiarism check done 

FYI - this pulled up on gitmemory?? 🤔

**Comment by justusmbuvi** [comment]
> [2120420-3bb0dc0c2c5d4e1f8480-1625592568154.pdf](https://github.com/section-engineering-education/engineering-education/files/6771739/2120420-3bb0dc0c2c5d4e1f8480-1625592568154.pdf)
> 
> Plagiarism check done
> 
> FYI - this pulled up on gitmemory??

@hectorkambow Is git memory the same as a published article? Should I change or modify it?

**Comment by justusmbuvi** [comment]
> Hey @justusmbuvi,
> 
> Welcome to EngEd community!
> 
> I'll be reviewing this PR.
> 
> Before I start the review, I would like a few changes to be done:
> 
> * Rename the directory `/Build-a-Twitch-Chatbot-in-Node.js/` to `/build-a-twitch-chatbot-in-nodejs/`
> * Add a hero image of dimensions 600x400. Use Unsplash to download images, and Pixlr to edit them (if needed)
> * Add authors page (headshot image + author bio). Have a look at `/contents/authors` to understand how its done. Or refer contribution guidelines to get started.
> 
> Feel free to ping me here or via Slack, in case of any questions.
> 
> Thanks.

Thanks, @srishilesh for the warm welcome! :100: 
I've added the hero image and the author... :rocket:

**Comment by justusmbuvi** [comment]
> Hey @justusmbuvi,
> 
> Welcome to EngEd community!
> 
> I'll be reviewing this PR.
> 
> Before I start the review, I would like a few changes to be done:
> 
> * Rename the directory `/Build-a-Twitch-Chatbot-in-Node.js/` to `/build-a-twitch-chatbot-in-nodejs/`
> * Add a hero image of dimensions 600x400. Use Unsplash to download images, and Pixlr to edit them (if needed)
> * Add authors page (headshot image + author bio). Have a look at `/contents/authors` to understand how its done. Or refer contribution guidelines to get started.
> 
> Feel free to ping me here or via Slack, in case of any questions.
> 
> Thanks.

Done!

File: content/articles/build-a-twitch-chatbot-in-nodejs/index.md [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+                // For example "PopCorn"
+                client.say(channel, `TwitchLit @${(mymessage.split(' ')[1] + '_' + mymessage.split(' ')[2])} TwitchLit  you have been UPVOTED by ${ tags.username }`);
+
+            } else if ((mymessage.split(' ')[0]).toLowerCase() === '!cheer' || 'cheers') {
+                console.log(`HSCheers @${(mymessage.split(' ')[1] + '_' + mymessage.split(' ')[2])} HSCheers you have been UPVOTED by ${ tags.username }`);
+            }
+            break;
+    }
+});
+```
```
Comment (srishilesh) [inline content/articles/build-a-twitch-chatbot-in-nodejs/index.md:386]: Break down this code snippet into smaller code snippets, and explain each of them in-detail.

File: content/articles/build-a-twitch-chatbot-in-nodejs/index.md [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+### Run it
+Run the code using:
+
+```bash
+npm run start
+```
+
+or
+
+```nodejs
```
Comment (srishilesh) [inline content/articles/build-a-twitch-chatbot-in-nodejs/index.md:399]:
```suggestion
```bash
```

File: content/articles/build-a-twitch-chatbot-in-nodejs/index.md [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+### Run the code to spy the chats
+Run this code using:
+
+```bash
+npm run start
+```
+
+or
+
+```nodejs
```
Comment (srishilesh) [inline content/articles/build-a-twitch-chatbot-in-nodejs/index.md:271]:
```suggestion
```bash
```

**Review by srishilesh — CHANGES_REQUESTED** [review]
@justusmbuvi,

I've done the preliminary review. Made a few changes.

I've also suggested for a few more changes, before we proceed for the tech review.

Please let me know when done.

Thanks.

File: content/articles/build-a-twitch-chatbot-in-nodejs/index.md (hunk 1 above)
Comment (justusmbuvi) [reply content/articles/build-a-twitch-chatbot-in-nodejs/index.md]: Ok

Comment (justusmbuvi) [reply content/articles/build-a-twitch-chatbot-in-nodejs/index.md]: Done!

Comment (justusmbuvi) [reply content/articles/build-a-twitch-chatbot-in-nodejs/index.md]: Committed the suggestions

**Review by srishilesh — APPROVED** [review]
@justusmbuvi 

Thanks for making the changes.

Moving for the final review. Make sure to go through my commits.

Thanks.

**Comment by hectorkambow** [comment]
@justusmbuvi 

Overall very nice job - please go over my changes and review what can be applied to future articles for improved reviews. 
Keep up the nice work. 

cc @srishilesh  please do the same - I've made some major changes for flow and format. 👍

**Pull request merged**